### PR TITLE
Added compliance test for whitespace in JSON literal.

### LINKED ISSUE #9: Test coverage needed for whitespace inside a backtick literal
cf. https://github.com/jmespath-community/jmespath.spec/pull/130

Example: ``` `1 ` ```

### PR DESCRIPTION
**Pull request by springcomp**
Fixes #9.

### CODE REVIEW

File: tests/literal.json [hunk 1]
```diff
@@ -78,6 +78,10 @@
                 "expression": "`9`",
                 "result": 9
             },
+            {
+              "expression": "`42 `",
+              "result": 42
+            },
```
Comment (gibson042) [inline tests/literal.json:84]: I think this is redundant with the "Literal with trailing whitespace" case in the "Literals" cases, and would recommend instead extending that section to include cases like
```json
        {
          "comment": "Literal with all valid leading and trailing JSON whitespace",
          "expression": "` \t\n\r{\"foo\": true} \t\n\r`",
          "result": {"foo": true}
        },
```
and a selection of non-JSON whitespace like
```json
        {
          "comment": "Literal with non-JSON whitespace U+000C FORM FEED",
          "expression": "`0\f`",
          "error": "syntax"
        },
        {
          "comment": "Literal with non-JSON whitespace U+0085 NEXT LINE",
          "expression": "`0\u0085`",
          "error": "syntax"
        },
        {
          "comment": "Literal with non-JSON whitespace U+00A0 NO-BREAK SPACE",
          "expression": "`0\u00A0`",
          "error": "syntax"
        },
        {
          "comment": "Literal with non-JSON whitespace U+1680 OGHAM SPACE MARK",
          "expression": "`0\u1680`",
          "error": "syntax"
        },
        {
          "comment": "Literal with non-JSON whitespace U+2028 LINE SEPARATOR",
          "expression": "`0\u2028`",
          "error": "syntax"
        },
        {
          "comment": "Literal with non-JSON whitespace U+3000 IDEOGRAPHIC SPACE",
          "expression": "`0\u3000`",
          "error": "syntax"
        },
```

Comment (springcomp) [reply tests/literal.json:84]: Thanks for the suggestion 👍.

This raises a point with regards to the Python implementation – and possibly the JavaScript implementation as well.
As JEP-12 has been introduced, those compliance tests effectively state that those constructs are now not standards-compliant. However, the Python implementation still supports the _legacy_ syntax and converts them to strings.

```patch
{
      "comment": "Literal with non-JSON whitespace U+1680 OGHAM SPACE MARK",
      "expression": "`0\u1680`",
-      "error": "syntax"
+     "result": "0\u1680"
},
```

I’m currently not too keen to break backwards compatibility as I’m trying hard to [get JMESPath Community adopted](https://github.com/aws/aws-cli/issues/7396) by [various stakeholders](https://github.com/Azure/azure-cli/issues/24542). What do you think?

Comment (gibson042) [reply tests/literal.json:84]: It looks like that issue was [anticipated in JEP-12](https://github.com/jmespath-community/jmespath.spec/blob/main/jep-012-raw-string-literals.md#impact) (emphasis mine):
> Implementations MAY choose to support the old syntax of allowing elided quotes in JSON literal expressions. If an implementation chooses this approach, the implementation SHOULD raise some kind of warning to the user to let them know of the deprecation and possible incompatibility with other JMESPath implementations.
> 
> In order to support this type of variance in JMESPath implementations, all of the JSON literal compliance test cases that involve elided quotes MUST be removed, and **test cases regarding failing on invalid unquoted JSON values MUST not be allowed in the compliance test unless placed in a JEP 12 specific test suite**, allowing implementations that support elided quotes in JSON literals to filter out the JEP 12 specific test cases.

The JEP-12 test suite was apparently never added, but now is the time—it should contain the above non-JSON whitespace cases.

> I’m currently not too keen to break backwards compatibility as I’m trying hard to [get JMESPath Community adopted](https://github.com/aws/aws-cli/issues/7396) by [various stakeholders](https://github.com/Azure/azure-cli/issues/24542). What do you think?

In that case, [jmespath.spec](https://github.com/jmespath-community/jmespath.spec) should be updated to document the mode switch with an optional (or required but associated with runtime syntax errors) `json-value =/ *( json-unescaped / json-escaped )` grammar production[^1] and a section explaining the consequences of an implementation decision regarding if or when to support it. But note also that implementation support for the legacy behavior isn't even internally consistent (e.g., https://github.com/jmespath/jmespath.js/issues/90 ).

[^1]: possibly abstracting `*( json-unescaped / json-escaped )` as something like `json-string-contents`

Comment (springcomp) [reply tests/literal.json]: I noticed that the James already [split compliance tests](https://github.com/jmespath-community/jmespath.py/tree/develop/tests/legacy) with this deprecation in mind.

So I can actually add those tests in the compliant file.

I also went ahead and [fully deprecated](https://github.com/jmespath-community/jmespath.py/blob/e79fdfe33bf0de9a193e740ed9a978c2e061b102/jmespath/lexer.py#L206-L215) pre-JEP12 JSON string literals while [adding a configuration flag](https://github.com/jmespath-community/jmespath.py/blob/e79fdfe33bf0de9a193e740ed9a978c2e061b102/jmespath/visitor.py#L77-L81) to maintain backwards compatibility.

File: tests/legacy/legacy_literal.json [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                "result": "foo\"quote"
+            },
+            {
+                "expression": "`✓`",
+                "result": "✓"
+            },
+            {
+                "comment": "Double quote in literal",
+                "expression": "`foo\\\"bar`",
+                "result": "foo\"bar"
```
Comment (gibson042) [inline tests/legacy/legacy_literal.json:24]: Isn't this test redundant with "Double quotes must be escaped."?

File: tests/literal.json [hunk 3]
```diff
@@ -132,6 +132,41 @@
...
           "expression": "foo.`\"bar\"`",
           "error": "syntax"
+        },
+        {
+          "comment": "Literal with all valid leading and trailing JSON whitespace",
+          "expression": "` \t\n\r{\"foo\": true} \t\n\r`",
+          "result": {"foo": true}
+        },
+        {
+          "comment": "Literal with non-JSON whitespace U+000C FORM FEED",
```
Comment (gibson042) [inline tests/literal.json:142]: Shouldn't the "literal with non-JSON whitespace" test cases be moved into their own file so implementations that support the legacy syntax can filter them out per [JEP-12: Impact](https://github.com/jmespath-community/jmespath.spec/blob/main/jep-012-raw-string-literals.md#impact) (emphasis mine)?
> In order to support this type of variance in JMESPath implementations, all of the JSON literal compliance test cases that involve elided quotes MUST be removed, and **test cases regarding failing on invalid unquoted JSON values MUST not be allowed in the compliance test unless placed in a JEP 12 specific test suite**, allowing implementations that support elided quotes in JSON literals to filter out the JEP 12 specific test cases.

I think we want to end up in a situation where any given test run includes either the legacy test cases or the syntax-error test cases but never both—or alternatively that the test cases are exactly the same but the expected results are configurable, e.g.
```json
        {
          "comment": "Literal with non-JSON whitespace U+000C FORM FEED",
          "expression": "`0\f`",
          "error": "syntax",
          "jep12LegacyResult": "0\f",
        },
```

Comment (springcomp) [reply tests/literal.json:142]: Hum that's a good point. However I did not want to dwell too long on expanding the existing legacy test suite that James already isolated.

The way I intend this folder organization is that common tests should be run with enforced standards compliance.

Tests in the legacy folder are intended for implementations that opt in to the old behaviour.

Comment (springcomp) [reply tests/literal.json:142]: That's how I plan the [upcoming changes to the Python implementation](https://github.com/jmespath-community/jmespath.py/tree/feature/vnext-preview/tests) to work.

See https://github.com/jmespath-community/jmespath.py/blob/69f3f9487e2360727e1252d21b24a367835464f9/tests/test_compliance.py#L74-L76

Comment (gibson042) [reply tests/literal.json:142]: If I understand that code correctly, the Python implementation testing is running all tests, but using legacy behavior for legacy test files and non-legacy behavior for all other test files. That will definitely work, but it means that any implementation which _doesn't_ have (configurable) non-legacy behavior must either skip the file completely (missing out on unrelated coverage) or add a custom approach for detecting failure to raise the expected syntax error on these specific (unmarked) test cases. Wouldn't it be better to instead reduce the burden on those implementations by isolating the test cases into their own file as suggested in JEP-12?

Comment (springcomp) [reply tests/literal.json]: Evidently, you were right.

File: tests/legacy/legacy_literal.json (hunk 2 above)
Comment (springcomp) [reply tests/legacy/legacy_literal.json]: Well it technically is. However, as the purpose of JEP-12 focused on the usage of mandatory double-quotes for surrounding test, I figure it does no harm to keep this text.

File: tests/legacy/legacy-literal.json [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      "type": "object"
+    },
+    "cases": [
+      {
+        "expression": "`foo`",
+        "result": "foo"
+      },
+      {
+        "expression": "`      foo`",
+        "result": "foo"
```
Comment (gibson042) [inline tests/legacy/legacy-literal.json:60]: Oh, interesting! I don't think there's any documentation regarding trimming whitespace in backtick literal legacy fallback behavior, and [JEP-12 Rationale](https://github.com/jmespath-community/jmespath.spec/blob/main/jep-012-raw-string-literals.md#rationale) suggests removal of leading and trailing whitespace (without a precise definition thereof) while implementations vary a bit but none act like that:
* [jmespath.py trims only leading whitespace](https://github.com/jmespath-community/jmespath.py/blob/bbe7300c60056f52413603cf3e2bcd0b6afeda3d/jmespath/lexer.py#L168), implicitly meaning Unicode property White_Space=Yes.
* [jmespath.js trims only leading whitespace](https://github.com/jmespath/jmespath.js/blob/f5f85697449b517f3f41969192a1d9367c4a0465/jmespath.js#L437), implicitly meaning Unicode property White_Space=Yes **OR** U+FEFF ZERO WIDTH NO-BREAK SPACE and **BUT NOT** U+0085 NEXT LINE.
* [jmespath.rb trims only leading whitespace](https://github.com/jmespath/jmespath.rb/blob/b2088367c4b6e72d6e24dbc02e2653a2b90a265c/lib/jmespath/lexer.rb#L337), implicitly meaning **ASCII** characters with Unicode property White_Space=Yes (i.e., U+0009 through U+000D plus U+0020 SPACE).
* [jmespath.php doesn't trim anything](https://github.com/jmespath/jmespath.php/blob/9b87907a81b87bc76d19a7fb2d61e61486ee9edb/src/Lexer.php#L434)

([go-jmespath](https://github.com/jmespath/go-jmespath/blob/b0104c826a24d7c202da8e1dbfb9ea4cfa35d774/parser.go#L321) and [jmespath.rs](https://github.com/jmespath/jmespath.rs/blob/93a173bbffe0c6012af9ce031d471fe3fbf244a9/jmespath/src/lexer.rs#L291-L294) do not support the fallback behavior AFAICT)

If this is codified in the compliance test suite, it should also be specified as part of https://github.com/jmespath-community/jmespath.spec/issues/133 .

**Review by gibson042 — APPROVED** [review]

File: tests/legacy/legacy_literal.json (hunk 2 above)
Comment (gibson042) [reply tests/legacy/legacy_literal.json]: I mean it's literally not testing anything interesting that isn't completely covered; the only difference is `quote` → `bar`.
```diff
-        "expression": "`foo\\\"quote`",
-        "result": "foo\"quote"
+        "expression": "`foo\\\"bar`",
+        "result": "foo\"bar"
```

File: tests/legacy/legacy-literal.json [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+      },
+      {
+        "expression": "`✓`",
+        "result": "✓"
+      },
+      {
+        "comment": "Double quote in literal",
+        "expression": "`foo\\\"bar`",
+        "result": "foo\"bar"
+      },
```
Comment (gibson042) [inline tests/legacy/legacy-literal.json:34]:
```suggestion
```

File: tests/legacy/legacy-literal.json (hunk 4 above)
Comment (springcomp) [reply tests/legacy/legacy-literal.json:55]: As I write [there](https://github.com/jmespath-community/jmespath.spec/issues/133#issuecomment-1314578465), I'm not sure I agree that the spec needs to be concerned with specifying legacy behaviour. However, I would be comfortable including this in any addendum made to JEP-12.

**Review by gibson042 — APPROVED** [review]
:+1:

**Pull request merged**